### PR TITLE
Add Velocity to PolyPIC P2G

### PR DESCRIPTION
**Pull request by sslattery**
By adding a separate particle velocity field to the PolyPIC P2G functions it allows for Lagrangian motion approximations to be used even if a non-velocity particle field is being mapped. Previously, if a non-velocity particle field was used this code would fail.

### CODE REVIEW

**Review by streeve — APPROVED** [review]
Looks good - I was able to at least build with what we discussed at the meeting

**Comment by kwitaechong** [comment]
I have some questions. 

1. I thought g2p operation is to update polypic velocity mode, Cpra as in the polypic tech note. Then, coeff(0) looks okay but coeff(1) and other coeffs look different from the definition of Cp2a and other Cpra

2. if we defined particle-wise local non-velocity field, c_p then don't we need to update this c_p in the g2p as we did for velocity mode, Cpra?

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -38,7 +38,7 @@ add_library(picasso ${SOURCES})
 
 target_link_libraries(picasso
   Kokkos::kokkos
-  Kokkos::kokkoskernels
+  kokkoskernels
```
[the remark this answers is not in the document]
Comment (sslattery) [reply src/CMakeLists.txt:41]: Oh this slipped in there from my local patch - I will remove

**Comment by sslattery** [comment]
> I have some questions.
> 
> 1. I thought g2p operation is to update polypic velocity mode, Cpra as in the polypic tech note. Then, coeff(0) looks okay but coeff(1) and other coeffs look different from the definition of Cp2a and other Cpra

We should really be able to update anything with this. The Cpra are just coefficients for individual scalar fields (e.g. velocity would have 3 scalar fields). I still need the velocity if I want to use equation 10 instead of equation 8 for the material motion approximation

> 2. if we defined particle-wise non-velocity field, c_p then don't we need to update this c_p in the g2p as we did for velocity mode, Cpra?

We also need to update Cpra for the non-velocity field. The idea is that you can use PolyPIC to advect any intensive field, not just the velocity/momentum.

**Comment by sslattery** [comment]
@kwitaechong Can you give me a review on this one too? I want to make sure you agree with the methodology.

**Comment by kwitaechong** [comment]
> > I have some questions.
> > 
> > 1. I thought g2p operation is to update polypic velocity mode, Cpra as in the polypic tech note. Then, coeff(0) looks okay but coeff(1) and other coeffs look different from the definition of Cp2a and other Cpra
> 
> We should really be able to update anything with this. The Cpra are just coefficients for individual scalar fields (e.g. velocity would have 3 scalar fields). I still need the velocity if I want to use equation 10 instead of equation 8 for the material motion approximation
> 
> > 1. if we defined particle-wise non-velocity field, c_p then don't we need to update this c_p in the g2p as we did for velocity mode, Cpra?
> 
> We also need to update Cpra for the non-velocity field. The idea is that you can use PolyPIC to advect any intensive field, not just the velocity/momentum.

**Review by kwitaechong — APPROVED** [review]
looks good

**Pull request merged**